src/params.rs

```diff
@@ -237,26 +277,55 @@ impl<PRFH: Hasher, MSGH: Hasher> Params<PRFH, MSGH> {
             hasher.write(p_seed.to_vec());
             hasher.write(vec![j as u8]);
             hasher.write(preimage);
-            let mut buf = vec![0u8; self.n as usize];
+            let mut buf = vec![0u8; PRFH::size()];
             hasher.sum(&mut buf);
-            curr_value.clone_from_slice(&buf);
+            curr_value.clone_from_slice(&buf[0..self.n]);
 
-            if generate {
-                chains[j as usize].copy_from_slice(&curr_value);
+            if let Some(ref mut chains) = maybe_chains {
+                let from = idx * self.n;
+                let to = (idx + 1) * self.n;
+                chains[j as usize + 1][from..to].copy_from_slice(&curr_value);
```

Comment:
we may need to add some length check before this kind of slice operations? for example

```rust
if chains.len() > end || chains.len() < start {
  return Err(OutOfRange)
}
```

Reply:
`chains` is declared in the calling function `compute_ladders` so the size should always be ok, but I can still add it if you'd like!